library/core/src/num/int_format.rs

```diff
@@ -0,0 +1,256 @@
+                offset -= 2;
+
+                let pair = (remain % 100) as usize;
+                remain /= 100;
+                buf.contents[offset + 0].write(DEC_DIGITS_LUT[pair * 2 + 0]);
+                buf.contents[offset + 1].write(DEC_DIGITS_LUT[pair * 2 + 1]);
+            }
+
+            // Format the last remaining digit, if any.
+            if remain != 0 || target_var == 0 {
+                // SAFETY: All of the decimals fit in buf, since it now is size-checked
+                // and the if condition ensures (at least) 1 more decimals.
+                unsafe { core::hint::assert_unchecked(offset >= 1) }
+                // SAFETY: The offset counts down from its initial value $BUF_SIZE
+                // without underflow due to the previous precondition.
+                unsafe { core::hint::assert_unchecked(offset <= $BUF_SIZE) }
+                offset -= 1;
+
+                // Either the compiler sees that remain < 10, or it prevents
+                // a boundary check up next.
+                let last = (remain & 15) as usize;
+                buf.contents[offset].write(DEC_DIGITS_LUT[last * 2 + 1]);
+                // not used: remain = 0;
+            }
+
+            offset
+        }
+    };
+}
+
+macro_rules! macro_impl_format_into {
+    (signed; $(($SignedT:ty, $UnsignedT:ty))*) => {
+        $(
+            #[unstable(feature = "int_format_into", issue = "138215")]
+            impl $SignedT {
+                /// Allows users to write an integer (in signed decimal format) into a variable `buf` of
+                /// type [`NumBuffer`] that is passed by the caller by mutable reference.
+                ///
+                /// This function panics if `buf` does not have enough size to store
+                /// the signed decimal version of the number.
+                ///
+                /// # Examples
+                /// ```
+                /// #![feature(int_format_into)]
+                /// use core::num::NumBuffer;
+                ///
+                #[doc = concat!("let n = -32", stringify!($SignedT), ";")]
+                /// let mut buf = NumBuffer::<3>::new();
+                /// assert_eq!(n.format_into(&mut buf), "-32");
+                ///
+                #[doc = concat!("let n2 = ", stringify!($SignedT::MIN), ";")]
+                /// let mut buf2 = NumBuffer::<40>::new();
+                #[doc = concat!("assert_eq!(n2.format_into(&mut buf2), ", stringify!($SignedT::MIN), ".to_string());")]
+                ///
+                #[doc = concat!("let n3 = ", stringify!($SignedT::MAX), ";")]
+                /// let mut buf3 = NumBuffer::<40>::new();
+                #[doc = concat!("assert_eq!(n3.format_into(&mut buf3), ", stringify!($SignedT::MAX), ".to_string());")]
+                /// ```
+                ///
+                pub fn format_into<const BUF_SIZE: usize>(self, buf: &mut crate::num::NumBuffer<BUF_SIZE>) -> &str {
```

Comment:
Ah, one more thing -- it would be very unfortunate to monomorphize all of this code for every distinct choice of buffer size (multiplied further by duplicate instantiations across separate codegen units). There's really not much benefit to be had by specializing the code for a particular buffer size (the only one I can think of is that a sufficiently clever compiler could elide the "is the size sufficient?" check if it's large enough for any value of the integer), so it's just a waste of binary size. As far as I know, no existing library on crates.io makes this choice either, and they're all quite performance conscious.